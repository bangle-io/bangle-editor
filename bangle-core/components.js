import { coreMarkPlugins, coreMarkSpec } from './marks/index';
import { coreNodeSpec, coreNodePlugins } from './nodes/index';
import * as history from './extensions/history';

export * from './nodes/index';
export * from './marks/index';
export * as history from './extensions/history';

export function coreSpec(opts = {}) {
  return [
    ...coreMarkSpec(opts.node),
    ...coreNodeSpec(opts.node),
    history.spec(opts.history),
  ];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts.node),
    ...coreNodePlugins(opts.node),
    history.plugins(opts.history),
  ];
}
