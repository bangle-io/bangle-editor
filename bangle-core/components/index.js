// marks
import * as bold from './bold';
import * as code from './code';
import * as italic from './italic';
import * as link from './link';
import * as strike from './strike';
import * as underline from './underline';

// nodes
import * as doc from './doc';
import * as paragraph from './paragraph';
import * as text from './text';
import * as blockquote from './blockquote';
import * as bulletList from './bullet-list';
import * as codeBlock from './code-block';
import * as hardBreak from './hard-break';
import * as heading from './heading';
import * as horizontalRule from './horizontal-rule';
import * as listItem from './list-item/list-item';
import * as orderedList from './ordered-list';
import * as todoItem from './todo-item';
import * as todoList from './todo-list';
import * as image from './image';

// components
import * as history from './history';

// marks
export * as bold from './bold';
export * as code from './code';
export * as italic from './italic';
export * as link from './link';
export * as strike from './strike';
export * as underline from './underline';

// nodes
export * as doc from './doc';
export * as paragraph from './paragraph';
export * as text from './text';
export * as blockquote from './blockquote';
export * as bulletList from './bullet-list';
export * as codeBlock from './code-block';
export * as hardBreak from './hard-break';
export * as heading from './heading';
export * as horizontalRule from './horizontal-rule';
export * as listItem from './list-item/list-item';
export * as orderedList from './ordered-list';
export * as todoItem from './todo-item';
export * as todoList from './todo-list';
export * as image from './image';

// components
export * as history from './history';

export function coreSpec(opts = {}) {
  return [...coreMarkSpec(opts), ...coreNodeSpec(opts)];
}

export function corePlugins(opts = {}) {
  return [
    ...coreMarkPlugins(opts),
    ...coreNodePlugins(opts),
    history.plugins(opts.history),
  ];
}

function coreMarkSpec(options = {}) {
  return [
    bold.spec(options.bold),
    code.spec(options.code),
    italic.spec(options.italic),
    strike.spec(options.strike),
    link.spec(options.link),
    underline.spec(options.underline),
  ];
}

function coreMarkPlugins(options = {}) {
  return [
    bold.plugins(options.bold),
    code.plugins(options.code),
    italic.plugins(options.italic),
    strike.plugins(options.strike),
    link.plugins(options.link),
    underline.plugins(options.underline),
  ];
}

function coreNodeSpec(options = {}) {
  return [
    doc.spec(options.doc),
    text.spec(options.text),
    paragraph.spec(options.paragraph),
    blockquote.spec(options.blockquote),
    bulletList.spec(options.bulletList),
    codeBlock.spec(options.codeBlock),
    hardBreak.spec(options.hardBreak),
    heading.spec(options.heading),
    horizontalRule.spec(options.horizontalRule),
    listItem.spec(options.listItem),
    orderedList.spec(options.orderedList),
    todoItem.spec(options.todoItem),
    todoList.spec(options.todoList),
    image.spec(options.image),
  ];
}

function coreNodePlugins(options = {}) {
  return [
    paragraph.plugins(options.paragraph),
    blockquote.plugins(options.blockquote),
    bulletList.plugins(options.bulletList),
    codeBlock.plugins(options.codeBlock),
    hardBreak.plugins(options.hardBreak),
    heading.plugins(options.heading),
    horizontalRule.plugins(options.horizontalRule),
    listItem.plugins(options.listItem),
    orderedList.plugins(options.orderedList),
    todoItem.plugins(options.todoItem),
    todoList.plugins(options.todoList),
    image.plugins(options.image),
  ];
}
