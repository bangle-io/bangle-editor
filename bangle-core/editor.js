import { EditorView } from 'prosemirror-view';
import { EditorState } from 'prosemirror-state';
import { pluginLoader } from './utils/plugin-loader';
import { isTestEnv } from './utils/environment';
import { DOMSerializer, DOMParser } from 'prosemirror-model';

export class BangleEditor {
  constructor(
    element,
    {
      specRegistry,
      plugins = [],
      viewOpts = {},
      editorProps,
      stateOpts = {},
      state = editorStateSetup({
        plugins,
        editorProps,
        specRegistry,
        stateOpts,
      }),
      focusOnInit = true,
    },
  ) {
    this._specRegistry = specRegistry;

    this._view = new EditorView(element, {
      state,
      dispatchTransaction(transaction) {
        const newState = this.state.apply(transaction);
        this.updateState(newState);
      },
      ...viewOpts,
      attributes: { class: 'bangle-editor' },
    });

    if (focusOnInit) {
      this.focusView();
    }
  }

  get view() {
    return this._view;
  }

  focusView() {
    const view = this._view;
    if (isTestEnv() || view.focused) {
      return;
    }
    view.focus();
  }

  toHTMLString() {
    const div = document.createElement('div');
    const fragment = DOMSerializer.fromSchema(
      this._specRegistry.schema,
    ).serializeFragment(this._view.state.doc.content);

    div.appendChild(fragment);

    return div.innerHTML;
  }

  destroy() {
    this._view.destroy();
  }
}

export function editorStateSetup({
  plugins = [],
  editorProps,
  specRegistry,
  stateOpts = {},
}) {
  const { doc, content, ...otherStateOpts } = stateOpts;

  const schema = specRegistry.schema;

  const state = EditorState.create({
    schema,
    doc: doc ? doc : createDocument({ schema, content: content }),
    ...otherStateOpts,
    plugins: pluginLoader(specRegistry, plugins, { editorProps }),
  });

  return state;
}

const createDocument = ({ schema, content, parseOptions }) => {
  const emptyDocument = {
    type: 'doc',
    content: [
      {
        type: 'paragraph',
      },
    ],
  };

  if (content == null) {
    return schema.nodeFromJSON(emptyDocument);
  }

  if (typeof content === 'object') {
    return schema.nodeFromJSON(content);
  }

  if (typeof content === 'string') {
    const element = document.createElement('div');
    element.innerHTML = content.trim();
    return DOMParser.fromSchema(schema).parse(element, parseOptions);
  }

  return false;
};
