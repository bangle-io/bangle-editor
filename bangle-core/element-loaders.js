import { InputRule } from 'prosemirror-inputrules';
import { Schema } from 'prosemirror-model';
import { inputRules, undoInputRule } from 'prosemirror-inputrules';

export function schemaLoader(specs) {
  let nodes = [];
  let marks = [];
  let topNode = 'doc';
  for (const spec of specs) {
    if (spec.type === 'node') {
      nodes.push([spec.name, spec.schema]);
    }
    if (spec.type === 'mark') {
      marks.push([spec.name, spec.schema]);
    }
    if (spec.topNode === true) {
      topNode = spec.name;
    }
  }

  return new Schema({
    topNode,
    nodes: Object.fromEntries(nodes),
    marks: Object.fromEntries(marks),
  });
}

export function loadInputRules(plugins) {
  let newPlugins = [];
  let match = [];
  plugins.forEach((plugin) => {
    if (plugin instanceof InputRule) {
      match.push(plugin);
      return;
    }
    newPlugins.push(plugin);
  });

  return [
    ...newPlugins,
    inputRules({
      rules: match,
    }),
  ];
}

export function loadPlugins(schema, plugins) {
  return plugins
    .flatMap((p) => {
      if (typeof p === 'function') {
        return p({ schema });
      }
      return p;
    })
    .filter(Boolean);
}
