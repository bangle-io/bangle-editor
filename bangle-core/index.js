import * as logging from './utils/logging';
import browser, { isChromeWithSelectionBug } from './utils/browser';
export * from './components/index';
export * from './editor';
export { logging, browser, isChromeWithSelectionBug };
export * from './pm-plugin';
export * from './node-view';
export * from './pm-model';
