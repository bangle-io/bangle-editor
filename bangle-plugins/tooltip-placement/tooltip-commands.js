const LOG = false;
let log = LOG
  ? console.log.bind(console, 'plugins/tooltip-commands')
  : () => {};

// Public commands
export const hideTooltip = (tooltipPlugin) => (state, dispatch) => {
  const pluginState = tooltipPlugin.getState(state);
  // Do not update an already hidden tooltip
  if (!pluginState.show) {
    log(tooltipPlugin, 'already hidden');
    return false;
  }

  log(tooltipPlugin, 'hiding tooltip');

  if (dispatch) {
    dispatch(
      state.tr
        .setMeta(tooltipPlugin, {
          show: false,
        })
        .setMeta('addToHistory', false),
    );
  }
  return true;
};

// Calling this command on a tooltip already showing will cause it to rerender i.e. update itself
export const showTooltip = (tooltipPlugin) => (state, dispatch) => {
  // We do not check when it is already in show state, to allow for the flexibility
  // of calling showTooltip multiple times as a way to signal updating of any downstream consumers
  // for example a tooltip can update its position
  log('tooltipPlugin, showing tooltip');
  if (dispatch) {
    dispatch(
      state.tr
        .setMeta(tooltipPlugin, {
          show: true,
        })
        .setMeta('addToHistory', false),
    );
  }
  return true;
};
