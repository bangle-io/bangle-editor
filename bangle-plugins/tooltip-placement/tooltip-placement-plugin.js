import { createPopper } from '@popperjs/core/lib/popper-lite';
import offset from '@popperjs/core/lib/modifiers/offset';
import preventOverflow from '@popperjs/core/lib/modifiers/preventOverflow';
import flip from '@popperjs/core/lib/modifiers/flip';
import arrow from '@popperjs/core/lib/modifiers/arrow';
import popperOffsets from '@popperjs/core/lib/modifiers/popperOffsets';
import { Plugin, PluginKey } from 'prosemirror-state';

const LOG = false;
let log = LOG
  ? console.log.bind(console, 'tooltip/tooltip-placement')
  : () => {};

/**
 * Dispatching show: true to the plugin will also update the tooltip position
 * Note: the state is interesting as it maintains the reference to the same state object if state is and was false
 * however any {show:true} dispatches will generate a new state object for easier reference checks. In a way
 * the state ref can be used to see if the position of tooltip has updated.
 *
 * To show a tooltip appendChild a div element to tooltipDOM with [data-popper-arrow] attribute
 *
 * @param {Object} options
 * @param {string} options.pluginName
 * @param {Element} options.tooltipDOM
 * @param {(view: any) => Element} options.getScrollContainerDOM
 * @param {(view: any, tooltipDOM: Element, scrollContainerDOM: Element) => {getBoundingClientRect: Function}} options.getReferenceElement
 * @param {string} options.placement
 * @param {(state: any, dispatch: any, view) => [number, number]} options.tooltipOffset
 * @param {(state: any, dispatch: any, view) => void} options.onUpdateTooltip - Called whenever tooltip is updated, will also be called when tooltip mounts for the first time
 * @param {(state: any, dispatch: any, view)  => void} options.onHideTooltip
 * @param {(view: any) => Array} options.customPopperModifiers
 * @param {(state: any) => Boolean} options.getInitialShowState
 * @param {Array} options.fallbackPlacement
 */
export function tooltipPlacementPlugin({
  pluginName = 'tooltipPlacementPlugin',
  tooltipDOM,
  getScrollContainerDOM,
  getReferenceElement,
  placement = 'top',
  tooltipOffset,
  onUpdateTooltip = (state, dispatch, view) => {},
  onHideTooltip = (state, dispatch, view) => {},
  customPopperModifiers,
  fallbackPlacements = ['bottom', 'top'],
  getInitialShowState = (state) => false,
}) {
  const key = new PluginKey(pluginName);
  tooltipDOM.setAttribute('data-tooltip-name', pluginName);

  const plugin = new Plugin({
    key: key,
    view: (view) => {
      return new TooltipPlacementView(view);
    },
    state: {
      init: (_, state) => {
        return {
          show: getInitialShowState(state),
        };
      },
      apply: (tr, pluginState) => {
        if (!tr.getMeta(key)) {
          return pluginState;
        }

        const newState = tr.getMeta(key);
        // Do not change object reference if 'show' was and is false
        if (newState.show === false && newState.show === pluginState.show) {
          return pluginState;
        }

        return newState;
      },
    },
  });

  class TooltipPlacementView {
    popperInstance = null;

    constructor(view) {
      this._view = view;
      this._tooltip = tooltipDOM;
      this._scrollContainerDOM = getScrollContainerDOM(view);
      // TODO should this be this plugins responsibility
      this._view.dom.parentNode.appendChild(this._tooltip);

      const pluginState = key.getState(view.state);
      // if the initial state is to show, setup the tooltip
      if (pluginState.show) {
        this._showTooltip();
        return;
      }
    }

    update(view, prevState) {
      const pluginState = key.getState(view.state);
      if (pluginState === key.getState(prevState)) {
        return;
      }
      if (pluginState.show) {
        log('calling updatetoolip ');

        onUpdateTooltip.call(this, view.state, view.dispatch, view);
        this._showTooltip();
      } else {
        log('calling hidetooltip');
        this._hideTooltip();
      }
    }

    destroy() {
      if (this.popperInstance) {
        this.popperInstance.destroy();
        this.popperInstance = null;
      }

      this._view.dom.parentNode.removeChild(this._tooltip);
    }

    _hideTooltip = () => {
      log('hiding');
      if (this.popperInstance) {
        this._tooltip.removeAttribute('data-show');
        this.popperInstance.destroy();
        this.popperInstance = null;
        onHideTooltip.call(
          this,
          this._view.state,
          this._view.dispatch,
          this._view,
        );
      }
    };

    _showTooltip = () => {
      this._tooltip.setAttribute('data-show', '');
      this._createPopperInstance(this._view);
      this.popperInstance.update();
    };

    _createPopperInstance(view) {
      if (this.popperInstance) {
        return;
      }

      const showTooltipArrow = this._tooltip.querySelector(
        '[data-popper-arrow]',
      );
      const defaultModifiers = [
        offset,
        preventOverflow,
        flip,
        {
          name: 'offset',
          options: {
            offset: (popperState) => {
              return tooltipOffset(popperState);
            },
          },
        },
        {
          name: 'flip',
          options: {
            fallbackPlacements,
            padding: 10,
          },
        },
        {
          name: 'preventOverflow',
          options: {
            boundary: this._scrollContainerDOM,
          },
        },
        popperOffsets,
        showTooltipArrow ? arrow : undefined,
        showTooltipArrow
          ? {
              name: 'arrow',
              options: {
                element: showTooltipArrow,
              },
            }
          : undefined,
      ].filter(Boolean);

      this.popperInstance = createPopper(
        getReferenceElement(view, this._tooltip, this._scrollContainerDOM),
        this._tooltip,
        {
          placement,
          modifiers: customPopperModifiers
            ? customPopperModifiers(
                view,
                this._tooltip,
                this._scrollContainerDOM,
                defaultModifiers,
              )
            : defaultModifiers,
        },
      );

      onUpdateTooltip.call(this, view.state, view.dispatch, view);
    }
  }

  return plugin;
}
