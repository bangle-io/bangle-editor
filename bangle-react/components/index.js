import * as dinos from './dinos/index';

export * as dinos from './dinos/index';

export const spec = (opts = {}) => {
  return [dinos.spec(opts.dinos)];
};

export const plugins = (opts = {}) => {
  return [dinos.plugins(opts.dinos)];
};

export const commands = {
  dinos: dinos.commands,
};
