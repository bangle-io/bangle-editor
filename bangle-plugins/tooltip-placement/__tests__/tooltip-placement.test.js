/**
 * @jest-environment jsdom
 */

/** @jsx psx */
import { psx } from 'bangle-core/test-helpers/index';

import { EditorState } from 'prosemirror-state';
import { tooltipPlacementPlugin } from '../tooltip-placement-plugin';
import { Editor as PMEditor, editorStateSetup } from 'bangle-core/editor';
import { hideTooltip, showTooltip } from '../tooltip-commands';
import { EditorView } from 'prosemirror-view';
import { createPopper } from '@popperjs/core/lib/popper-lite';
import { corePlugins, coreSpec } from 'bangle-core/components';
import { schemaLoader } from 'bangle-core/element-loaders';

jest.mock('@popperjs/core/lib/popper-lite', () => {
  return {
    createPopper: jest.fn(() => ({
      update: jest.fn(),
      destroy: jest.fn(),
    })),
  };
});

const setupPlugin = (opts = {}) => {
  const plugin = tooltipPlacementPlugin({
    tooltipDOM: document.createElement('div'),
    getScrollContainerDOM: () => {
      return document.createElement('div');
    },
    getReferenceElement: () => {
      return {
        getBoundingClientRect: () => {
          return {
            width: 0,
            height: 0,
            top: 0,
            right: 0,
            bottom: 0,
            left: 0,
          };
        },
      };
    },
    ...opts,
  });
  return plugin;
};

const setupEditorState = (plugin) => {
  const editorSpec = [...coreSpec()];
  const plugins = [...corePlugins(), plugin];

  return editorStateSetup({
    plugins,
    editorSpec,
    doc: (<doc>
      <para>hello world</para>
    </doc>)(schemaLoader(editorSpec)),
  });
};

const setupEditorView = ({
  state,
  viewDOM = document.createElement('div'),
}) => {
  return new EditorView(viewDOM, {
    state,
  });
};

describe('Tooltip state', () => {
  test('Default is state false', async () => {
    const plugin = setupPlugin();
    let state = setupEditorState(plugin);

    expect(plugin.getState(state)).toMatchInlineSnapshot(`
      Object {
        "show": false,
      }
    `);
  });

  test('Obeys initial showState', async () => {
    const plugin = setupPlugin({
      getInitialShowState: () => true,
    });
    let state = setupEditorState(plugin);

    expect(plugin.getState(state)).toMatchInlineSnapshot(`
      Object {
        "show": true,
      }
    `);
  });

  test('hideTooltip command works', async () => {
    const plugin = setupPlugin({
      getInitialShowState: () => true,
    });
    let state = setupEditorState(plugin);

    expect(plugin.getState(state).show).toBe(true);

    hideTooltip(plugin)(state, (tr) => {
      state = state.apply(tr);
    });

    expect(plugin.getState(state).show).toBe(false);
  });

  test("hideTooltip doesn't dispatch if already hidden", async () => {
    const plugin = setupPlugin({});
    let state = setupEditorState(plugin);

    const dispatch = jest.fn((tr) => {
      state = state.apply(tr);
    });

    hideTooltip(plugin)(state, dispatch);

    expect(dispatch).toBeCalledTimes(0);
    expect(plugin.getState(state).show).toBe(false);
  });

  test('showTooltip command works', async () => {
    const plugin = setupPlugin({
      getInitialShowState: () => false,
    });
    let state = setupEditorState(plugin);

    showTooltip(plugin)(state, (tr) => {
      state = state.apply(tr);
    });

    expect(plugin.getState(state).show).toBe(true);
  });

  test('showTooltip command creates new state even if already showing', async () => {
    const plugin = setupPlugin({
      getInitialShowState: () => true,
    });

    let state = setupEditorState(plugin);

    const pluginState = plugin.getState(state);
    expect(pluginState).toMatchInlineSnapshot(`
      Object {
        "show": true,
      }
    `);

    showTooltip(plugin)(state, (tr) => {
      state = state.apply(tr);
    });
    expect(pluginState).toEqual(plugin.getState(state));
    expect(pluginState).not.toBe(plugin.getState(state));
  });
});

describe('plugin view', () => {
  test('passed options correctly', async () => {
    const tooltipDOM = document.createElement('div');
    const referenceElement = jest.fn();
    const getReferenceElement = jest.fn(() => referenceElement);
    const scrollContainerDOM = document.createElement('div');
    const plugin = setupPlugin({
      fallbackPlacements: ['pos-a', 'pos-b'],
      getInitialShowState: (state) => true,
      getReferenceElement,
      getScrollContainerDOM: () => scrollContainerDOM,
      placement: 'test-placement',
      tooltipDOM,
    });
    const view = setupEditorView({ state: setupEditorState(plugin) });

    expect(createPopper).toBeCalledTimes(1);
    expect(createPopper).nthCalledWith(1, referenceElement, tooltipDOM, {
      placement: 'test-placement',
      modifiers: expect.any(Array),
    });

    expect(createPopper.mock.calls[0][2].modifiers).toMatchSnapshot();
    expect(getReferenceElement).toBeCalledTimes(1);
    expect(getReferenceElement).nthCalledWith(
      1,
      view,
      tooltipDOM,
      scrollContainerDOM,
    );
  });

  test('Mounts ', async () => {
    let mockPopperInstance;
    const onUpdateTooltip = jest.fn(function () {
      mockPopperInstance = this.popperInstance;
    });
    const onHideTooltip = jest.fn();
    const tooltipDOM = document.createElement('div');
    const viewDOM = document.createElement('div');
    const plugin = setupPlugin({
      onUpdateTooltip,
      onHideTooltip,
      tooltipDOM,
    });
    const view = setupEditorView({ state: setupEditorState(plugin), viewDOM });

    expect(plugin.getState(view.state).show).toBe(false);

    expect(createPopper).toBeCalledTimes(0);
    expect(viewDOM.contains(tooltipDOM)).toBe(true);

    showTooltip(plugin)(view.state, view.dispatch);

    expect(tooltipDOM.hasAttribute('data-show')).toBe(true);

    expect(plugin.getState(view.state).show).toBe(true);

    expect(createPopper).toBeCalledTimes(1);
    expect(mockPopperInstance.update).toBeCalledTimes(1);
    // one for creating popper and one for showing
    // i know its redundant but its not expected for this
    // callback to be pure
    expect(onUpdateTooltip).toBeCalledTimes(2);
    expect(onUpdateTooltip).nthCalledWith(1, view.state, view.dispatch, view);
    expect(onUpdateTooltip).nthCalledWith(2, view.state, view.dispatch, view);
    expect(onHideTooltip).toBeCalledTimes(0);
  });

  test('hide tooltip', async () => {
    let mockPopperInstance;
    const onUpdateTooltip = jest.fn(function () {
      mockPopperInstance = this.popperInstance;
    });
    const tooltipDOM = document.createElement('div');

    const onHideTooltip = jest.fn();
    const plugin = setupPlugin({
      onUpdateTooltip,
      onHideTooltip,
      tooltipDOM,
    });
    const view = setupEditorView({ state: setupEditorState(plugin) });

    expect(plugin.getState(view.state).show).toBe(false);

    expect(createPopper).toBeCalledTimes(0);
    showTooltip(plugin)(view.state, view.dispatch);
    expect(plugin.getState(view.state).show).toBe(true);
    expect(mockPopperInstance.destroy).toBeCalledTimes(0);
    expect(tooltipDOM.hasAttribute('data-show')).toBe(true);

    hideTooltip(plugin)(view.state, view.dispatch);

    expect(tooltipDOM.hasAttribute('data-show')).toBe(false);

    expect(onHideTooltip).toBeCalledTimes(1);
    expect(onHideTooltip).nthCalledWith(1, view.state, view.dispatch, view);
    expect(mockPopperInstance.destroy).toBeCalledTimes(1);
  });

  test('view destroy', async () => {
    const onUpdateTooltip = jest.fn(function () {});
    const tooltipDOM = document.createElement('div');

    const onHideTooltip = jest.fn();
    const plugin = setupPlugin({
      onUpdateTooltip,
      onHideTooltip,
      tooltipDOM,
    });

    const viewDOM = document.createElement('div');

    const view = setupEditorView({ state: setupEditorState(plugin), viewDOM });

    expect(plugin.getState(view.state).show).toBe(false);

    showTooltip(plugin)(view.state, view.dispatch);

    view.destroy();

    expect(viewDOM.contains(tooltipDOM)).toBe(false);
  });
});
