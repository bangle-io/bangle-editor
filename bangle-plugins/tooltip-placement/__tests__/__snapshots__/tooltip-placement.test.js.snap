// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`plugin view passed options correctly 1`] = `
Array [
  Object {
    "enabled": true,
    "fn": [Function],
    "name": "offset",
    "phase": "main",
    "requires": Array [
      "popperOffsets",
    ],
  },
  Object {
    "enabled": true,
    "fn": [Function],
    "name": "preventOverflow",
    "phase": "main",
    "requiresIfExists": Array [
      "offset",
    ],
  },
  Object {
    "data": Object {
      "_skip": false,
    },
    "enabled": true,
    "fn": [Function],
    "name": "flip",
    "phase": "main",
    "requiresIfExists": Array [
      "offset",
    ],
  },
  Object {
    "name": "offset",
    "options": Object {
      "offset": [Function],
    },
  },
  Object {
    "name": "flip",
    "options": Object {
      "fallbackPlacements": Array [
        "pos-a",
        "pos-b",
      ],
      "padding": 10,
    },
  },
  Object {
    "name": "preventOverflow",
    "options": Object {
      "boundary": <div />,
    },
  },
  Object {
    "data": Object {},
    "enabled": true,
    "fn": [Function],
    "name": "popperOffsets",
    "phase": "read",
  },
]
`;
